### DISCUSSION TITLE
Trigger a camera (from NIDAQ analog input)

### DISCUSSION THREAD

**Discussion by camille-lab** [bonsai-rx/bonsai#1352] [Q&A] [answered]
Hello all!

**What I want:**
I am trying to trigger video acquisition from a camera. 
I want the camera recording to start when a trigger comes in, record for x seconds, save that video and wait for next trigger (next behavior trial).

I have been able to do the manual trigger-record-save part using the keydown node, so that part is OK.
For the automatic trigger part:
I think ideally I would read a trigger on the GPIO of the camera and that's it. But I have a Basler camera and it seems that Bonsai cannot access that type of metadata from Basler cameras (see also [this post ](https://groups.google.com/g/bonsai-users/c/6Xb2hJ_GYak/m/j2GCNG8xFQAJ))

**Workaround:**
I have been trying to read the trigger as an analog input signal from a USB NIDAQ.
Flow (attached): I extract the desired channel, I convert it to be able to threshold it to go back to a digital signal, and then 'distinctUnitlChanged' to fire the record-save node only when receiving a trigger. (I would also need to filter out the change from 1->0). However, bonsai keeps on saving videos for some reasons (even without any trigger).
I think it's a format type issue (NIDAQ output a OpenCV.Net.Mat).
The format type is also the reason why I cannot use the 'greaterThan' node, that should do the job just fine.

Any suggestions people? For fixing the workflow or another idea?
Thanks a lot!

Camille

[NIDAQ_2cams_debugging.zip](https://github.com/bonsai-rx/bonsai/files/11440645/NIDAQ_2cams_debugging.zip)

**Comment by jbscherrer** [comment]
This is the workflow structure that I use for triggering video saving. If the undesired video saving is happening when you start the workflow, the issue may be that you need a Skip (1) node at the output of your trigger signal, because your trigger line will generate an output on startup.

![image](https://github.com/bonsai-rx/bonsai/assets/39060975/2e7629e6-2941-4cbe-9e48-54ab608c6655)

Within selectmany:
![image](https://github.com/bonsai-rx/bonsai/assets/39060975/c98f8138-744c-401f-9044-2bd4298af376)

**Comment by camille-lab** [comment]
Hi Joe,
thanks for sharing your workflow. 
I suppose your 'CameraCapture' node is inside the 'video_buffer' one? And what is inside the 'RisingEdge'?
I think the problem that I am facing is to turn the data from the NIDAQ into something these nodes can deal with (even after convert scale)...perhaps because of the metadata embedded with the data? do you have any idea how to do that?

**Comment by jbscherrer** [comment]
That's right, there is a VideoCaptureDevice in video_buffer. RisingEdge is simply: 
![image](https://github.com/bonsai-rx/bonsai/assets/39060975/d7cd691c-6137-4bcc-98d6-3f2993351630)
which only propagates values that are true. Because the input is DistinctUntilChanged, the only values that propagate correspond to transitions from low to high.

The AnalogInput nodes from the NIDAQ output batches of samples in OpenCV.Net.Mat format of size equal to the node property BufferSize. These buffers arrive at a rate given by SampleRate/BufferSize. I would set these parameters to give you the temporal precision you need, and then perform a matrix operation to convert each matrix into a single value (i.e. find the max value within each matrix, then report just that number for each buffer that arrives). You could also use a digital input on the DAQ and a DigitalInput node.

**Comment by camille-lab** [comment]
thanks again for your reply, Joe!

It would be easier with a digital input indeed, but the DigitalInput node does not seem to work... Am I correct?

I am not familiar with the OpenCV.Net.Mat format. I have been trying to use the 'ConvertToMat' node but it doesn't do the job. 
Would you mind walking me through the steps, from the nidaq format to whatever is convenient in bonsai? (i suppose, double?), or pointing me to help that is intelligible for newbie?

Thanks a lot

**Reply by jbscherrer** [reply, marked as answer]
I would feed the OpenCV.Net.Mat into an `Average` node, and feed that into a `MemberSelector` that pulls out `Val0` (right click on the `Average` node, select Output>Val0). That will give you an output of type `double`.

**Comment by camille-lab** [comment]
Thanks Joe, it seems to be working!
